google/cloud/documentai_toolbox/wrappers/document_revision.py

```diff
@@ -0,0 +1,502 @@
+    r"""Modifies the entities using the providence field in the entities.
+
+    Args:
+        entities (List[Entity]):
+            A list of documentai.Document.Entity.
+
+    Returns:
+        Tuple[list,list]:
+            modified entities, Entity history.
+    """
+    entities_array = []
+    history = []
+    for e in entities:
+        if (
+            e.provenance.type_ == _OP_TYPE.ADD
+            or e.provenance.type_ == _OP_TYPE.OPERATION_TYPE_UNSPECIFIED
+        ):
+            entities_array.append(Entity(documentai_entity=e))
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": e.type_,
+                    "original_text": e.mention_text,
+                }
+            )
+        elif e.provenance.type_ == _OP_TYPE.REMOVE:
+            entity = entities_array.pop(int(e.id) - 1)
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": entity.type_,
+                    "original_text": entity.mention_text,
+                }
+            )
+            del entity
+        elif e.provenance.type_ == _OP_TYPE.REPLACE:
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": e.type_,
+                    "original_text": e.mention_text,
+                    "replace_type": e.type_,
+                    "replace_text": e.mention_text,
+                }
+            )
+            entities_array[int(e.id) - 1] = e
+    return entities_array, history
+
+
+def _get_level(current_revision: "DocumentWithRevisions") -> int:
+    r"""Returns the level of the current revision in the revision tree.
+
+    Args:
+        current_revision (DocumentWithRevisions):
+            The current revision.
+
+    Returns:
+        int:
+            The level of the current revision in the revision tree.
+    """
+    try:
+        level = 0
+        p = current_revision.parent
+        if p.revision_id:
+            while p:
+                p = p.parent
+                level += 1
+        return level
+    except Exception:
+        return 0
+
+
+def _print_child_tree(
+    current_revision: "DocumentWithRevisions",
+    doc: "DocumentWithRevisions",
+    seen: List[str],
+):
+    r"""Prints the revision tree of a child revision.
+
+    Args:
+        None.
+
+    Returns:
+        None.
+    """
+    tab = "  " * _get_level(doc)
+
+    if doc.children:
+        if current_revision.revision_id == doc.revision_id:
+            print(tab + "└──>", end="")
```

Comment:
I am a little uneasy about maintaining code that draws ASCII art.  Can we explore libraries that do this for us?

Reply:
I would agree. This could also be something to explore for the print_document_tree in gcs_utilities.py